torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+    // Wrap the new block under a trivial thread-idx
+    //   for t in 0..1: // threadIdx
+    //     if (t < 1):
+    //       new_block
+    VarHandle t("t", kInt);
+    ExprHandle t_lt_1 = CompareSelect::make(t, 1, CompareSelectOperation::kLT);
+    // TODO: move "if (t < 1)" to threadIdx generation
+    Cond* masked_block = Cond::make(t_lt_1, new_block, nullptr);
+    LoopOptions thread_idx_opt;
+    // TODO: the added trivial threadIdx needs to match the kernel threadIdx
+    // dimensions
+    thread_idx_opt.set_gpu_thread_index(0);
+    For* trivial_loop = For::make(t, 0, 1, masked_block, thread_idx_opt);
+    return trivial_loop;
+  }
+
+  Stmt* mutate(const For* v) override {
+    if (v->loop_options().is_gpu_block_index()) {
+      gpu_blocks_.push_back(v);
+      need_rewrite_ = false;
+    } else if (v->loop_options().is_gpu_thread_index()) {
+      gpu_threads_.push_back(v);
+    }
+
+    Stmt* new_for = IRMutator::mutate(v);
+
+    if (v->loop_options().is_gpu_block_index()) {
+      gpu_blocks_.pop_back();
+      need_rewrite_ = false;
+    } else if (v->loop_options().is_gpu_thread_index()) {
+      gpu_threads_.pop_back();
+    }
+
+    return new_for;
+  }
+
+  Stmt* mutate(const Block* v) override {
+    std::list<Stmt*> old_stmts = v->stmts();
+    std::vector<bool> need_rewrites(old_stmts.size());
+    std::vector<Stmt*> new_stmts(old_stmts.size());
+    ;
+    int index = 0;
+    for (auto old_stmt : old_stmts) {
+      need_rewrite_ = false;
+      Stmt* new_stmt = old_stmt->accept_mutator(this);
+      need_rewrites[index] = need_rewrite_;
+      new_stmts[index] = new_stmt;
+      index++;
+    }
+
+    bool any_need_fix = false;
+    bool all_need_fix = true;
+    for (auto need_fix : need_rewrites) {
+      if (need_fix) {
+        any_need_fix = true;
+      } else {
+        all_need_fix = false;
+      }
+    }
+
+    need_rewrite_ = false;
+    // If nothing needs fix, return as it is
+    if (!any_need_fix) {
+      return (Stmt*)v;
+    }
+
+    // If all needs fix, check if parent is a GPU block
+    if (all_need_fix) {
+      Stmt* parent = v->get_parent();
+      For* loop_parent = dynamic_cast<For*>(parent);
+      if (loop_parent) {
+        if (loop_parent->loop_options().is_gpu_block_index()) {
+          Stmt* new_block = rewrite(new_stmts);
+          return new_block;
+        }
+      }
+      need_rewrite_ = true;
+      return (Stmt*)v;
+    }
+
+    // if some needs fix, rewrites the consecutive parts
+    int start = 0;
+    int count = new_stmts.size();
+    std::vector<Stmt*> rewrite_stmts;
+    while (start < count) {
+      while (start < count && !need_rewrites[start]) {
+        rewrite_stmts.push_back(Stmt::clone(new_stmts[start]));
+        start++;
+      }
+      int stop = start + 1;
+      while (stop < count && need_rewrites[stop]) {
+        stop++;
+      }
+
+      // Rewrite the stmts from [start, stop)
+      std::vector<Stmt*> stmts_block(
```

Comment:
Nit: the name `stmts_block` is slightly misleading as it's not a `Block` (even though it's very similar by its meaning).